greenmail-core/src/main/java/com/icegreen/greenmail/Managers.java

```diff
@@ -4,21 +4,104 @@
  */
 package com.icegreen.greenmail;
 
+import java.lang.reflect.Constructor;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Type;
+
+import com.icegreen.greenmail.configuration.GreenMailConfiguration;
+import com.icegreen.greenmail.filestore.UncheckedFileStoreException;
 import com.icegreen.greenmail.imap.ImapHostManager;
 import com.icegreen.greenmail.imap.ImapHostManagerImpl;
 import com.icegreen.greenmail.smtp.SmtpManager;
-import com.icegreen.greenmail.store.InMemoryStore;
+import com.icegreen.greenmail.store.Store;
 import com.icegreen.greenmail.user.UserManager;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 /**
  * @author Wael Chatila
  * @version $Id: $
  * @since Jan 27, 2006
  */
 public class Managers {
-    private ImapHostManager imapHostManager = new ImapHostManagerImpl(new InMemoryStore());
-    private UserManager userManager = new UserManager(imapHostManager);
-    private SmtpManager smtpManager = new SmtpManager(imapHostManager, userManager);
+    final Logger log = LoggerFactory.getLogger(Managers.class);
+
+    private ImapHostManager imapHostManager = null;
+    private UserManager userManager = null;
+    private SmtpManager smtpManager = null;
+    private Store storeToUse = null;
+
+    /**
+     * Public constructor wihtout Startup Configuration
+     *
+     * @deprecated Please use the constructor with a valid Startup Configuration
+     */
+    public Managers() {
+        this(new GreenMailConfiguration());
+    }
+
+    /**
+     * Public constructor with startupConfiguration
+     *
+     * @param startupConfig - the startup configuration
+     */
+    public Managers(GreenMailConfiguration startupConfig) {
+        log.info("Starting managers with the following startup configuration:");
+        startupConfig.logConfiguration();
+
+        // Depending on the startup confifguration, construct the correct Store class using Reflection:

```

Comment:
While I wouldn't go so far as to say that reflection is generally bad, I think it's a bit of an overkill in this scenario. You could just go with a factory that returns the correct store based on the config using a switch statement.